HelpersTask560_Create_dockerized_executable_for_gh_issue_labels

sandeepthalapanane opened the pull request:
Addressing issue #560

Created a Dockerized executable containing the external packages for `dev_scripts_helpers/github/sync_gh_issue_labels.py`, following the example 
- https://github.com/causify-ai/helpers/blob/master/dev_scripts_helpers/notebooks/dockerized_extract_notebook_images.py.

Please let me know if I should implement any test cases.

sonniki commented:
@heanhsok feel free to take the lead on review

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,262 @@
+# #############################################################################
+
+
+class Label:
```
The dockerized executable is meant to be a just wrapper around the actual utility script, so the implementation logic should be kept there.

Feel free to look a few samples from the explanation doc in the issue.

sandeepthalapanane replied:
The dockerized executable [dockerized_extract_notebook_images.py](https://github.com/causify-ai/helpers/blob/master/dev_scripts_helpers/notebooks/dockerized_extract_notebook_images.py) 
contains the utility class  _NotebookImageExtractor. So I just replicated based on this example.

gpsaggese replied:
I understand the confusion here and I have probably not been consistent myself.

@heanhsok as per our chat, I'm ok using the convention of calling scripts `xyz.py` (which contain only the docker part and the call to the actual script), and then we call the "actual" script doing the work `dockerized_xyz.py`

The reason is:
- We are asking to do something (`xyz`) and the fact that it's done with Docker is an implementation detail
- The innermost script can be called `dockerized_...` in the sense that it needs to run inside a docker container.

Can you update the doc with this info and then we clean it up?

heanhsok commented on dev_scripts_helpers/github/sync_gh_issue_labels.py:
```diff
@@ -176,90 +71,141 @@ def _parse() -> argparse.ArgumentParser:
+    hdbg.init_logger(
+        verbosity=args.log_level, use_exec_path=True, force_white=False
+    )
+    _run_dockerized_sync_gh_issue_labels(
```
This function call go inside the dockerized_* file so people can run

```bash
> ./dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py \
    --input_file ./dev_scripts_helpers/github/labels/gh_issues_labels.yml \
    --owner causify-ai \
    --repo tutorials \
    --token_env_var GITHUB_TOKEN \
    --dry_run \
    --backup
```

sandeepthalapanane replied:
As said above, the `_run_dockerized_*` function call is in the [extract_notebook_images.py](https://github.com/causify-ai/helpers/blob/master/dev_scripts_helpers/notebooks/extract_notebook_images.py) script. Even in another doc example, the `_run_dockerized_llm_transform` is called in the [llm_transform.py](https://github.com/causify-ai/helpers/blob/master/dev_scripts_helpers/llms/llm_transform.py). I am slightly confused with the example and the things you mentioned. Should I look at the examples or implement the things you mentioned?

gpsaggese replied:
@heanhsok if we use the convention above, do you agree that this is ok then?

heanhsok reviewed (changes requested):
Good start. Let's do a bit of structural change first.

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,262 @@
+"""
+The script is designed to synchronize GitHub issue labels from a label
+inventory manifest file. It requires certain dependencies to be present (e.g.,
+`pygithub`) and thus it is executed within a Docker container.
```
... as a dockerized executable.

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,262 @@
+inventory manifest file. It requires certain dependencies to be present (e.g.,
+`pygithub`) and thus it is executed within a Docker container.
+
+To use this script, you need to provide the input file, GitHub
```
Let's put some examples of command lines, or better because of DRY let's say "The command lines are the same as sync_gh_...`

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,262 @@
+_WARNING = "\033[33mWARNING\033[0m"
+
+# TODO(*): outsourceable. Convert to dockerized executable.
+try:
+    import github
+except ModuleNotFoundError:
+    _module = "pygithub"
+    print(_WARNING + f": Can't find {_module}: continuing")
```
I doubt the script can continue and do anything without the import, but ok

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,262 @@
+def _main(parser: argparse.ArgumentParser) -> None:
+    args = parser.parse_args()
+    hdbg.init_logger(verbosity=args.log_level, use_exec_path=True)
+    # Load labels from label inventory manifest file.
```
I'd add a TODO to split this into functions. Cursor does it easily (e.g., "Replace this code with a function and call it," and then you move the function out).

It's P1 since this is not an important script, but nonetheless principles are principles

gpsaggese commented on dev_scripts_helpers/github/sync_gh_issue_labels.py:
```diff
@@ -176,90 +71,141 @@ def _parse() -> argparse.ArgumentParser:
+    )
+    # Build the command.
+    git_root = hgit.find_git_root()
+    docker_executable = "sync_gh_issue_labels.py"
```
Uhm... should this be "dockerized_..."?

sandeepthalapanane replied:
Yes, it should be dockerized

heanhsok commented on dev_scripts_helpers/github/sync_gh_issue_labels.py:
```diff
@@ -176,90 +71,141 @@ def _parse() -> argparse.ArgumentParser:
+    # Build the container image, if needed.
+    container_image = "tmp.sync_gh_issue_labels"
+    dockerfile = r"""
+    FROM python:3.10-slim
```
I think it needs `git` to be installed as well

Feel to do a dry run to manually test this

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_issue_labels.py:
```diff
@@ -0,0 +1,267 @@
+    # Save the labels if backup is enabled.
+    if args.backup:
+        git_root_dir = hgit.get_client_root(False)
+        file_name = f"tmp.labels.{args.owner}.{args.repo}.yaml"
```
Could you pls also check if this file is correctly output in the right place on the host system (when things are running from docker)?

sandeepthalapanane replied:
Yes, the backup file is correctly saved to my system's `helpers` repository root.

sandeepthalapanane commented:
> @sandeepthalapanane once you've addressed the comments, let's also do a dry run as a way to manually test the script, since we don't have a unit test yet

I performed the dry run, and below is the command I ran and the output. 

```bash
> python ./dev_scripts_helpers/github/sync_gh_issue_labels.py  --input_file ./dev_scripts_helpers/github/labels/gh_issues_labels.yml  --owner causify-ai  --repo tutorials --token_env_var GITHUB_TOKEN     --dry_run  --backup  --no_interactive 
```

```bash
19:55:30 - INFO  hdbg.py init_logger:1013                               Saving log to file '/home/sandeep/src/helpers1/dev_scripts_helpers/github/sync_gh_issue_labels.py.log'
19:55:30 - INFO  hdbg.py init_logger:1018                               > cmd='./dev_scripts_helpers/github/sync_gh_issue_labels.py --input_file ./dev_scripts_helpers/github/labels/gh_issues_labels.yml --owner causify-ai --repo tutorials --token_env_var GITHUB_TOKEN --dry_run --backup --no_interactive'
19:55:32 - INFO  sync_gh_issue_labels.py _main:211                      Label synchronization completed!
```

gpsaggese commented:
LGTM. @heanhsok
1) Should we move to testing this?
2) Is it even worth it to test this?
3) Should we use some mocking + the idea in https://github.com/causify-ai/helpers/issues/615#issue-3018607302 to check the sequence of commands?

Feel free to create a followup. I'd pass it to @sandeepthalapanane since he knows the code now

heanhsok commented:
LGTM.

There's not much to test here in this script, as we're mainly using the library code to perform actions. Basic testing to see if they can be run and produce the correct output should suffice. I'd put it as P1.

Feel free to file a follow-up @sandeepthalapanane
